## ID3 Tree
##### Jimmy Hu

In [6]:
import numpy as np
import pandas as pd
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split

In [2]:
df = pd.read_csv('final result.csv')
df = df[df.columns[2:-1].to_list()]

X = df[df.columns[1:]].values
y = df['Mortality (1= death)'].values

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.33, random_state=42)

In [12]:
dtc = DecisionTreeClassifier(criterion='entropy', ccp_alpha=0.001)
dtc.fit(X_train, y_train)

DecisionTreeClassifier(ccp_alpha=0.001, class_weight=None, criterion='entropy',
                       max_depth=None, max_features=None, max_leaf_nodes=None,
                       min_impurity_decrease=0.0, min_impurity_split=None,
                       min_samples_leaf=1, min_samples_split=2,
                       min_weight_fraction_leaf=0.0, presort='deprecated',
                       random_state=None, splitter='best')

In [50]:
feature = dtc.feature_importances_ > 0
d = dict(zip(list(df.columns[1:].values[feature]),
             list(dtc.feature_importances_[feature].round(3))))
ranking = dict(sorted(d.items(), key=lambda x: x[1], reverse=True))

In [13]:
dtc.get_depth()

6

In [14]:
dtc.score(X_test, y_test)

0.9665437539347064

In [49]:
print('Feature Importance:')
for i in ranking.keys():
    print(ranking[i], i)

Feature Importance:
0.548 CVASC_Cardiac_B
0.238 CVASC_Other_Nos_B
0.094 ENDOC_MET_Diabetes
0.05 CVASC_Heart_Rhythm_A
0.03 CHEST_Airway_Lungs_A
0.023 UROLG_GEN_Other_Nos_B
0.017 Age_75_99


考虑到feature全是binary的，我用最简单的ID3 Tree来切分整个样本，并设置一个Minimal Cost-Complexity Pruning的threshold防止overfitting。最终一个depth为6的ID3 Tree的预测准确率能达到96.7%。

**事实上只有很少的几个feature导致了死亡**